utilities/checkpoint/checkpoint_impl.cc

```diff
@@ -253,6 +253,10 @@ Status CheckpointImpl::CreateCustomCheckpoint(
     // We cannot get min_log_num before calling the GetLiveFiles() for the
     // first time, because if we do that, all the logs files will be included,
     // far more than needed.
+    if (!db_->GetIntProperty(DB::Properties::kMinLogNumberToKeep,
+                             &min_log_num)) {
```

Comment:
it's tricky, the above existing comments explained the reasoning.
Without this, CheckpointTest.CurrentFileModifiedWhileCheckpointing2PC will fail.

Reply:
The above comment explains why GetLiveFiles() is called twice, not min_log_num is fetched twice, unless I missed anything.

Reply:
You are right. I proposed a new solution, PTAL.

Reply:
Since `GetLiveFiles` will do flush if flush is true, getting min_log_num before `GetLiveFiles` will cause the min_log_num to be unnecessarily small, but getting it after `GetLiveFiles` has the problem described in the PR's description. 
So I think the proper way is to get `min_log_num` and `manifest_size` while holding the mutex.

Reply:
I think it is a good idea. We probably need to preserve the current interface and add a new one to DB. A lot of places need to change then.

Since it is getting a little bit deeper, maybe @pdillinger can take a quick look since he worked on checkpoint/backup recently and might have more opinions on it.

Reply:
@cheng-chang what would be the problem if min_log_num is smaller than needed? We just hardlink more files that might not be strictly needed, right? It shouldn't be a big problem?

Reply:
Yes, we just hardlink more files. But since it's a hardlink, it means those WALs will never be removed from disk as long as the checkpoint exists. Not sure whether this will be a concern.

Reply:
@siying  what's your opinion on the current solution of getting the live files twice?

Reply:
> what would be the problem if min_log_num is smaller than needed? We just hardlink more files that might not be strictly needed, right? It shouldn't be a big problem?

I thought WALs are copied.

> I think it is a good idea. We probably need to preserve the current interface and add a new one to DB. A lot of places need to change then.

Agree it is a good idea. Is the DB interface supposed to be extensible? If not I think we can add a pointer-to-int argument that defaults to nullptr, preserving API compatibility. If it is supposed to be extensible, we can instead add an overload that takes an additional pointer-to-int argument. Then the old overload can call the new overload passing `nullptr` for that argument. I may be missing the main difficulty with that approach.